VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
+
+This method registers a new endpoint on the network:
+
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
+6. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
+
+Unregister a storage node from the network at the specified date:
+
+1. `didAddress`: The DID that is to be removed from the network
+2. `unregisterDatetime`: The unix timestamp of when the storage node should no longer be available for selection. Must be at least 28 days in the future.
+3. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.
+
+## getNodeByAddress(didAddress: string): StorageNode
+
+## getNodeByEndpoint(endpointUri: string): StorageNode
+
+## addDatacenter(name: string, countryCode: string, regionCode: string, lat: int, long: int) ownerOnly
+
+Add a data center to the network. `id` will be autoincremented.
+
+Returns `int` the unique ID of the created data center
+
+## removeDatacenter(id: string) ownerOnly
+
+Remove a data center by `id`. Will only remove the data center if there are no storage nodes using that datacenter.
+
+## getDatacenters(ids: string[]): Datacenter[]
```

Comment:
For Storage Nodes to register themselves, they will have to pass the Id of their datacenter.
This `getDatacenters(ids)` taking a list of ids as parameter, doesn't make it easy for them to find their exact datacenter?
Maybe a `getAllDatacenters()` or a `getDatacentersByCountry(countryCode)`

Reply:
Good point.

Perhaps we add:

- `getDatacentersByCountry(countryCode: string): DataCenter[]`
- `getDatacentersByRegion(regionCode: string): DataCenter[]`

Thoughts @itstar?

Reply:
Ok, will add these functions.

Reply:
@ITStar10 Please add them to this draft spec document as well